File: app/src/main/java/io/github/zwieback/familyfinance/business/iconics/activity/IconicsActivity.java
package io.github.zwieback.familyfinance.business.iconics.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.Menu;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;

import com.annimon.stream.IntPair;
import com.annimon.stream.Optional;
import com.annimon.stream.Stream;
import com.jakewharton.rxbinding3.appcompat.RxSearchView;
import com.mikepenz.iconics.Iconics;
import com.mikepenz.iconics.typeface.IIcon;
import com.mikepenz.iconics.typeface.ITypeface;
import com.mikepenz.iconics.typeface.library.community.material.CommunityMaterial;
import com.mikepenz.materialdrawer.Drawer;
import com.mikepenz.materialdrawer.DrawerBuilder;
import com.mikepenz.materialdrawer.holder.BadgeStyle;
import com.mikepenz.materialdrawer.holder.StringHolder;
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem;
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeUnit;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.business.iconics.fragment.IconicsFragment;
import io.github.zwieback.familyfinance.business.iconics.listener.OnIconSelectListener;
import io.github.zwieback.familyfinance.core.activity.ActivityWrapper;
import io.github.zwieback.familyfinance.core.drawer.DrawerListener;
import io.github.zwieback.familyfinance.util.NumberUtils;
import io.reactivex.android.schedulers.AndroidSchedulers;

public class IconicsActivity extends ActivityWrapper implements OnIconSelectListener {

    public static final String OUTPUT_ICON_NAME = "iconName";

    private List<ITypeface> mFonts;
    private String mCurrentSearch;
    private Drawer mDrawer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Toolbar toolbar = findViewById(R.id.toolbar);
        mFonts = getAndSortFonts();
        List<IDrawerItem<?>> items = buildDrawerItems(mFonts);
        int mIdentifierCmd = determineIdentifierOfCommunityMaterial(mFonts);
        mDrawer = buildDrawer(toolbar, mFonts, items, mIdentifierCmd);
    }

    @Override
    protected void setupContentView() {
        setContentView(R.layout.activity_iconics);
    }

    @Override
    protected int getTitleStringId() {
        return 0;
    }

    @Override
    protected boolean isDisplayHomeAsUpEnabled() {
        return false;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_entity_search, menu);
        SearchView searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
        setupSearchView(searchView);
        return super.onCreateOptionsMenu(menu);
    }

    private void setupSearchView(SearchView searchView) {
        addDisposable(
                RxSearchView.queryTextChanges(searchView)
                        .debounce(NumberUtils.UI_DEBOUNCE_TIMEOUT, TimeUnit.MILLISECONDS,
                                AndroidSchedulers.mainThread())
                        .subscribe(searchName -> {
                            mCurrentSearch = searchName.toString();
                            if (mDrawer != null) {
                                Stream.of(mFonts)
                                        .forEachIndexed((index, font) -> {
                                            long foundCount = findMatchedIconCount(mCurrentSearch, font);
                                            mDrawer.updateBadge(index, new StringHolder(foundCount + ""));
                                        });
                            }
                            searchInFragment(mCurrentSearch);
                        })
        );
    }

    @Override
    public void onIconSelected(@NonNull String iconName) {
        Intent resultIntent = new Intent();
        resultIntent.putExtra(OUTPUT_ICON_NAME, iconName);
        setResult(Activity.RESULT_OK, resultIntent);
        finish();
    }

    private long findMatchedIconCount(String searchName, ITypeface font) {
        String searchInLowerCase = searchName.toLowerCase();
        return Stream.ofNullable(font.getIcons())
                .filter(icon -> icon.toLowerCase().contains(searchInLowerCase))
                .count();
    }

    private List<ITypeface> getAndSortFonts() {
        List<ITypeface> fonts = new ArrayList<>(Iconics.getRegisteredFonts(this));
        Collections.sort(fonts, (typeface1, typeface2) ->
                typeface1.getFontName().compareTo(typeface2.getFontName()));
        return fonts;
    }

    private List<IDrawerItem<?>> buildDrawerItems(List<ITypeface> fonts) {
        List<IDrawerItem<?>> items = new ArrayList<>(fonts.size());
        int count = 0;
        for (ITypeface font : fonts) {
            String description =
                    TextUtils.isEmpty(font.getAuthor())
                            ? font.getVersion()
                            : font.getVersion() + " - " + font.getAuthor();

            IDrawerItem drawerItem = new PrimaryDrawerItem()
                    .withName(font.getFontName())
                    .withBadge(String.valueOf(font.getIcons().size()))
                    .withDescription(description)
                    .withBadgeStyle(new BadgeStyle().withColorRes(R.color.md_grey_200))
                    .withIcon(extractRandomIcon(font))
                    .withIdentifier(count);

            items.add(drawerItem);
            count++;
        }
        return items;
    }

    private IIcon extractRandomIcon(ITypeface typeface) {
        int random = new Random().nextInt(typeface.getIcons().size());
        Optional<String> randomIconName = Stream.of(typeface.getIcons())
                .findIndexed((index, icon) -> index == random)
                .map(IntPair::getSecond);
        return typeface.getIcon(randomIconName.orElse(typeface.getIcons().iterator().next()));
    }

    private int determineIdentifierOfCommunityMaterial(List<ITypeface> fonts) {
        String communityMaterialFontName = CommunityMaterial.INSTANCE.getFontName();

        Optional<Integer> communityMaterialIdentifier = Stream.of(fonts)
                .findIndexed((index, font) -> communityMaterialFontName.equals(font.getFontName()))
                .map(IntPair::getFirst);
        return communityMaterialIdentifier.orElse(0);
    }

    private Drawer buildDrawer(Toolbar toolbar,
                               List<ITypeface> fonts,
                               List<IDrawerItem<?>> items,
                               int selectedItem) {
        return new DrawerBuilder()
                .withActivity(this)
                .withToolbar(toolbar)
                .withDrawerItems(items)
                .withOnDrawerItemClickListener((view, position, iDrawerItem) -> {
                    loadIcons(fonts.get(position).getFontName());
                    if (getSupportActionBar() != null) {
                        getSupportActionBar().setTitle(fonts.get(position).getFontName());
                    }
                    return false;
                })
                .withOnDrawerListener(new DrawerListener(this))
                .withFireOnInitialOnClick(true)
                .withSelectedItem(selectedItem)
                .build();
    }

    private void loadIcons(@NonNull String fontName) {
        IconicsFragment iconicsFragment = IconicsFragment.newInstance(fontName);
        iconicsFragment.onSearch(mCurrentSearch);
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.iconics_fragment, iconicsFragment)
                .commit();
    }

    private void searchInFragment(String searchName) {
        IconicsFragment fragment = findFragment();
        if (fragment != null) {
            fragment.onSearch(searchName);
        }
    }

    private IconicsFragment findFragment() {
        return (IconicsFragment) getSupportFragmentManager().findFragmentById(R.id.iconics_fragment);
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/iconics/activity/IconicsActivity.kt
package io.github.zwieback.familyfinance.business.iconics.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import com.jakewharton.rxbinding3.appcompat.queryTextChanges
import com.mikepenz.iconics.Iconics
import com.mikepenz.iconics.typeface.IIcon
import com.mikepenz.iconics.typeface.ITypeface
import com.mikepenz.iconics.typeface.library.community.material.CommunityMaterial
import com.mikepenz.materialdrawer.Drawer
import com.mikepenz.materialdrawer.DrawerBuilder
import com.mikepenz.materialdrawer.holder.BadgeStyle
import com.mikepenz.materialdrawer.holder.StringHolder
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.business.iconics.fragment.IconicsFragment
import io.github.zwieback.familyfinance.business.iconics.listener.OnIconSelectListener
import io.github.zwieback.familyfinance.constant.UiConstants
import io.github.zwieback.familyfinance.core.activity.ActivityWrapper
import io.github.zwieback.familyfinance.core.drawer.DrawerListener
import io.reactivex.android.schedulers.AndroidSchedulers
import java.util.Locale
import java.util.concurrent.TimeUnit
import kotlin.Comparator
import kotlin.collections.ArrayList
import kotlin.math.max
import kotlin.random.Random

class IconicsActivity : ActivityWrapper(), OnIconSelectListener {

    private lateinit var fonts: List<ITypeface>
    private lateinit var drawer: Drawer
    private var currentSearch: String? = null

    override val titleStringId: Int
        get() = 0

    override val isDisplayHomeAsUpEnabled: Boolean
        get() = false

    private val registeredSortedFonts: List<ITypeface>
        get() = Iconics.getRegisteredFonts(this)
            .sortedWith(Comparator { typeface1, typeface2 ->
                typeface1.fontName.compareTo(typeface2.fontName)
            })

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        fonts = registeredSortedFonts
        val items = buildDrawerItems(fonts)
        val identifierCmd = determineIdentifierOfCommunityMaterial(fonts)
        drawer = buildDrawer(toolbar, fonts, items, identifierCmd)
    }

    override fun setupContentView() {
        setContentView(R.layout.activity_iconics)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_entity_search, menu)
        val searchView = menu.findItem(R.id.action_search).actionView as SearchView
        setupSearchView(searchView)
        return super.onCreateOptionsMenu(menu)
    }

    private fun setupSearchView(searchView: SearchView) {
        addDisposable(
            searchView.queryTextChanges()
                .debounce(
                    UiConstants.UI_DEBOUNCE_TIMEOUT,
                    TimeUnit.MILLISECONDS,
                    AndroidSchedulers.mainThread()
                )
                .subscribe { searchName ->
                    currentSearch = searchName.toString()
                    currentSearch?.let { currentSearch ->
                        fonts
                            .forEachIndexed { index, font ->
                                val foundCount = findMatchedIconCount(currentSearch, font)
                                drawer.updateBadge(index.toLong(), StringHolder("$foundCount"))
                            }
                    }
                    searchInFragment(currentSearch)
                }
        )
    }

    override fun onIconSelected(iconName: String) {
        val resultIntent = Intent()
            .putExtra(OUTPUT_ICON_NAME, iconName)
        setResult(Activity.RESULT_OK, resultIntent)
        finish()
    }

    private fun findMatchedIconCount(searchName: String, font: ITypeface): Int {
        val searchInLowerCase = searchName.toLowerCase(Locale.getDefault())
        return font.icons
            .asSequence()
            .filter { icon -> icon.toLowerCase(Locale.getDefault()).contains(searchInLowerCase) }
            .count()
    }

    private fun buildDrawerItems(fonts: List<ITypeface>): List<IDrawerItem<*>> {
        val items = ArrayList<IDrawerItem<*>>(fonts.size)
        for ((index, font) in fonts.withIndex()) {
            val description = if (font.author.isEmpty()) {
                font.version
            } else {
                "${font.version} - ${font.author}"
            }
            val drawerItem = PrimaryDrawerItem()
                .withName(font.fontName)
                .withBadge(font.icons.size.toString())
                .withDescription(description)
                .withBadgeStyle(BadgeStyle().withColorRes(R.color.md_grey_200))
                .withIcon(extractRandomIcon(font))
                .withIdentifier(index.toLong())

            items.add(drawerItem)
        }
        return items
    }

    private fun extractRandomIcon(typeface: ITypeface): IIcon {
        val randomIndex = Random.Default.nextInt(typeface.icons.size)
        val randomIconName = typeface.icons
            .asSequence()
            .filterIndexed { index, _ -> index == randomIndex }
            .firstOrNull()
            ?: typeface.icons.iterator().next()
        return typeface.getIcon(randomIconName)
    }

    private fun determineIdentifierOfCommunityMaterial(fonts: List<ITypeface>): Int {
        val communityMaterialFontName = CommunityMaterial.fontName
        val communityMaterialIdentifier = fonts
            .indexOfFirst { font -> communityMaterialFontName == font.fontName }
        return max(communityMaterialIdentifier, 0)
    }

    private fun buildDrawer(
        toolbar: Toolbar,
        fonts: List<ITypeface>,
        items: List<IDrawerItem<*>>,
        selectedItemIdentifier: Int
    ): Drawer {
        return DrawerBuilder()
            .withActivity(this)
            .withToolbar(toolbar)
            .withDrawerItems(items)
            .withOnDrawerItemClickListener(object : Drawer.OnDrawerItemClickListener {
                override fun onItemClick(
                    view: View?,
                    position: Int,
                    drawerItem: IDrawerItem<*>
                ): Boolean {
                    val fontName = fonts[position].fontName
                    supportActionBar?.title = fontName
                    loadIcons(fontName)
                    return false
                }
            })
            .withOnDrawerListener(DrawerListener(this))
            .withFireOnInitialOnClick(true)
            .withSelectedItem(selectedItemIdentifier.toLong())
            .build()
    }

    private fun loadIcons(fontName: String) {
        val iconicsFragment = IconicsFragment.newInstance(fontName)
        iconicsFragment.onSearch(currentSearch)
        supportFragmentManager.beginTransaction()
            .replace(R.id.iconics_fragment, iconicsFragment)
            .commit()
    }

    private fun searchInFragment(searchName: String?) {
        findFragment()?.onSearch(searchName)
    }

    private fun findFragment(): IconicsFragment? {
        return supportFragmentManager.findFragmentById(R.id.iconics_fragment) as IconicsFragment?
    }

    companion object {
        const val OUTPUT_ICON_NAME = "iconName"
    }
}
